cpp/include/cudf/detail/distinct_hash_join.cuh

```diff
@@ -85,22 +86,42 @@ struct hasher_adapter {
 template <cudf::has_nested HasNested>
 struct distinct_hash_join {
  private:
-  /// Device row equal type
-  using d_equal_type = cudf::experimental::row::equality::strong_index_comparator_adapter<
-    cudf::experimental::row::equality::device_row_comparator<HasNested == cudf::has_nested::YES,
-                                                             cudf::nullate::DYNAMIC>>;
+  using row_comparator = cudf::experimental::row::equality::device_row_comparator<
+    true,
+    cudf::nullate::DYNAMIC,
+    cudf::experimental::row::equality::nan_equal_physical_equality_comparator,
+    cudf::experimental::type_identity_t>;
+
+  using row_comparator_no_nested = cudf::experimental::row::equality::device_row_comparator<
+    false,
+    cudf::nullate::DYNAMIC,
+    cudf::experimental::row::equality::nan_equal_physical_equality_comparator,
+    cudf::experimental::dispatch_void_if_nested_t>;
+
+  using row_comparator_no_compound = cudf::experimental::row::equality::device_row_comparator<
+    false,
+    cudf::nullate::DYNAMIC,
+    cudf::experimental::row::equality::nan_equal_physical_equality_comparator,
+    cudf::experimental::dispatch_void_if_compound_t>;
+
   using hasher              = hasher_adapter<thrust::identity<hash_value_type>>;
   using probing_scheme_type = cuco::linear_probing<1, hasher>;
   using cuco_storage_type   = cuco::storage<1>;
 
   /// Hash table type
-  using hash_table_type = cuco::static_set<cuco::pair<hash_value_type, rhs_index_type>,
-                                           cuco::extent<size_type>,
-                                           cuda::thread_scope_device,
-                                           comparator_adapter<d_equal_type>,
-                                           probing_scheme_type,
-                                           cudf::detail::cuco_allocator<char>,
-                                           cuco_storage_type>;
+  template <typename Comparator>
+  using static_set_with_comparator = cuco::static_set<
+    cuco::pair<hash_value_type, rhs_index_type>,
+    cuco::extent<size_type>,
+    cuda::thread_scope_device,
+    comparator_adapter<
+      cudf::experimental::row::equality::strong_index_comparator_adapter<Comparator>>,
+    probing_scheme_type,
```

Comment:
question: probing scheme type also wraps row hasher type. Do we want to dispatch row hasher in this case as well?

Reply:
It looks like we only perform a identity function on the hash value stored in the hash table, so I think this wouldn't bring benefit
```
using hasher              = hasher_adapter<thrust::identity<hash_value_type>>;
  using probing_scheme_type = cuco::linear_probing<1, hasher>;
  using cuco_storage_type   = cuco::storage<1>;

  /// Hash table type
  template <typename Comparator>
  using static_set_with_comparator = cuco::static_set<
    cuco::pair<hash_value_type, rhs_index_type>,
    cuco::extent<size_type>,
    cuda::thread_scope_device,
    comparator_adapter<
      cudf::experimental::row::equality::strong_index_comparator_adapter<Comparator>>,
    probing_scheme_type,
    cudf::detail::cuco_allocator<char>,
    cuco_storage_type>;
```